File: scripts/front-proxy.js
// 可选首页前置层（Node 应用用）：
// / 和 /index.html 返回 index.html，其余所有请求（含 WebSocket）转发给你的应用。
// 你的应用以后台方式运行在 PORT+10 上。
const { spawn } = require('child_process');
const http = require('http');
const net = require('net');
const fs = require('fs');
const path = require('path');

const HERE = __dirname;
const PORT = parseInt(process.env.PORT || '3000', 10);
const BACK = PORT + 10;
const HTML = fs.readFileSync(path.join(HERE, 'index.html'));

// 启动真正的应用
// stdio 必须用 'ignore'：部分应用会劫持/改写自己的 stdout，继承控制台 fd
// 在 unikernel 环境下会把进程卡死（CPU 0%、无日志），别改回 inherit
spawn(process.execPath, ['index.js'], {
  cwd: HERE,
  env: { ...process.env, PORT: String(BACK) },
  stdio: 'ignore',
});

function proxy(req, res) {
  const p = http.request(
    { host: '127.0.0.1', port: BACK, path: req.url, method: req.method, headers: req.headers },
    (r) => { res.writeHead(r.statusCode, r.headers); r.pipe(res); }
  );
  p.on('error', () => { try { res.writeHead(502); res.end('bad gateway'); } catch (e) {} });
  req.pipe(p);
}

const server = http.createServer((req, res) => {
  const u = (req.url || '/').split('?')[0];
  if (u === '/' || u === '/index.html') {
    res.writeHead(200, {
      'Content-Type': 'text/html; charset=utf-8',
      'Content-Length': HTML.length,
      'Cache-Control': 'no-cache',
    });
    return res.end(req.method === 'HEAD' ? undefined : HTML);
  }
  proxy(req, res);
});

// WebSocket 透传
server.on('upgrade', (req, socket, head) => {
  const up = net.connect(BACK, '127.0.0.1', () => {
    const keep = {};
    for (let i = 0; i < req.rawHeaders.length; i += 2) {
      if (/^connection$|^upgrade$/i.test(req.rawHeaders[i])) keep[req.rawHeaders[i].toLowerCase()] = req.rawHeaders[i + 1];
    }
    let raw = `${req.method} ${req.url} HTTP/1.1\r\nHost: 127.0.0.1:${BACK}\r\n`;
    for (let i = 0; i < req.rawHeaders.length; i += 2) {
      const k = req.rawHeaders[i];
      if (/^host$/i.test(k)) continue;
      raw += `${k}: ${req.rawHeaders[i + 1]}\r\n`;
    }
    raw += '\r\n';
    up.write(raw);
    if (head && head.length) up.write(head);
    up.pipe(socket);
    socket.pipe(up);
  });
  up.on('error', () => socket.destroy());
  socket.on('error', () => up.destroy());
});

server.listen(PORT, () => console.log(`front :${PORT} -> app :${BACK}`));

File: scripts/front-proxy.py
#!/usr/bin/env python3
"""可选首页前置层（Python 应用用）：
/ 和 /index.html 返回 index.html，其余 HTTP 请求转发给你的应用(:PORT+10)。
注意：不支持 WebSocket 透传（Node 版支持）。
"""
import http.server
import os
import subprocess
import sys
import urllib.error
import urllib.request

HERE = os.path.dirname(os.path.abspath(__file__))
PORT = int(os.environ.get('PORT', '3000'))
BACK = PORT + 10
HTML = open(os.path.join(HERE, 'index.html'), 'rb').read()
SKIP_REQ = {'host', 'content-length', 'connection', 'accept-encoding'}
SKIP_RESP = {'connection', 'transfer-encoding'}

# 启动真正的应用（入口：环境变量 PY_ENTRY 优先，否则按 main.py/app.py/index.py 探测）
ENTRY = os.environ.get('PY_ENTRY', '')
if ENTRY not in ('main.py', 'app.py', 'index.py'):
    ENTRY = next((f for f in ('main.py', 'app.py', 'index.py')
                  if os.path.exists(os.path.join(HERE, f))), 'main.py')
# stdout/stderr 必须 DEVNULL：继承控制台 fd 在 unikernel 下可能卡死进程
subprocess.Popen([sys.executable, '-u', ENTRY],
                 cwd=HERE, env={**os.environ, 'PORT': str(BACK)},
                 stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)


class Handler(http.server.BaseHTTPRequestHandler):
    protocol_version = 'HTTP/1.1'

    def _page(self):
        self.send_response(200)
        self.send_header('Content-Type', 'text/html; charset=utf-8')
        self.send_header('Content-Length', str(len(HTML)))
        self.send_header('Cache-Control', 'no-cache')
        self.end_headers()
        if self.command != 'HEAD':
            self.wfile.write(HTML)

    def _fwd(self):
        body = None
        n = self.headers.get('Content-Length')
        if n:
            body = self.rfile.read(int(n))
        req = urllib.request.Request('http://127.0.0.1:%d%s' % (BACK, self.path),
                                     data=body, method=self.command)
        for k, v in self.headers.items():
            if k.lower() not in SKIP_REQ:
                req.add_header(k, v)
        try:
            with urllib.request.urlopen(req, timeout=60) as r:
                data = r.read()
                status, headers = r.status, r.headers.items()
        except urllib.error.HTTPError as e:
            data = e.read()
            status, headers = e.code, e.headers.items()
        except Exception:
            data, status, headers = b'bad gateway', 502, []
        try:
            self.send_response(status)
            sent_len = False
            for k, v in headers:
                if k.lower() in SKIP_RESP or k.lower() == 'content-length':
                    continue
                self.send_header(k, v)
            self.send_header('Content-Length', str(len(data)))
            self.end_headers()
            if self.command != 'HEAD':
                self.wfile.write(data)
        except Exception:
            pass

    def route(self):
        u = (self.path or '/').split('?')[0]
        if u in ('/', '/index.html'):
            self._page()
        else:
            self._fwd()

    do_GET = do_POST = do_PUT = do_DELETE = do_PATCH = do_HEAD = route

    def log_message(self, *a):  # 静音访问日志
        pass


if __name__ == '__main__':
    print('front :%d -> app :%d' % (PORT, BACK))
    http.server.ThreadingHTTPServer(('0.0.0.0', PORT), Handler).serve_forever()
